File: day-02/part2.py
from util.run import run
from util.direction.direction import Direction, InvalidDirection


def solution(moves: list[tuple[Direction, int]]):
    aim = 0
    depth = 0
    horizontal = 0

    for (dir, scalar) in moves:
        if dir == Direction.DOWN:
            aim += scalar
        elif dir == Direction.UP:
            aim -= scalar
        elif dir == Direction.FORWARD:
            horizontal += scalar
            depth += aim * scalar
        else:
            raise InvalidDirection()

    return horizontal * depth


if __name__ == "__main__":
    run(solution)

File: day-02/part2.go
package main

import (
	"github.com/dawidl022/advent-of-code-2021/day-02/util"
	"github.com/dawidl022/advent-of-code-2021/day-02/util/direction"
)

func solution2(moves []direction.Move) int {
	aim := 0
	depth := 0
	horizontal := 0

	for _, move := range moves {
		switch move.Dir {
		case direction.DOWN:
			aim += move.Scalar
		case direction.UP:
			aim -= move.Scalar
		case direction.FORWARD:
			horizontal += move.Scalar
			depth += aim * move.Scalar
		}
	}
	return horizontal * depth
}

func main() {
	util.Run(solution2)
}
